src/dataflow/components/dataflow-program.tsx

```diff
@@ -160,13 +164,43 @@ export class DataflowProgram extends BaseComponent<IProps, IState> {
         if (change) {
           (async () => {
             await engine.abort();
-            await engine.process(editor.toJSON());
+            await engine.process(this.programEditor.toJSON());
           })();
         }
       });
 
-      editor.view.resize();
-      (editor as any).trigger("process");
+      this.programEditor.view.resize();
+      (this.programEditor as any).trigger("process");
+
+      this.intervalHandle = setInterval(this.heartBeat.bind(this), HEARTBEAT_INTERVAL);
+
     })();
   }
+
+  public componentWillUnmount() {
+    clearInterval(this.intervalHandle);
+  }
+
+  private heartBeat() {
+    this.programEditor.nodes.forEach((n: Node) => {
+      if (n.data.hasOwnProperty("nodeValue")) {
+        const val: any = n.data.nodeValue;
+        if (n.data.recentValues) {
+          const values: any = n.data.recentValues;
+          if (values.length > MAX_NODE_VALUES) {
+            values.shift();
+          }
+          values.push(val);
+          n.data.recentValues = values;
+        } else {
+          const values: number[] = [val];
+          n.data.recentValues = values;
+        }
+        const plotControl = n.controls.get("plot") as PlotControl;
+        if (plotControl) {
+          (plotControl as any).update();
+        }
+      }
+    });
+  }
```

Comment:
It seems awkward to have the program editor reach down into the implementation of the nodes. Would it be possible to have the nodes be responsible for their own updates? Then the heartbeat would just be:
```
this.programEditor.nodes.forEach((n: Node) => { n.update(); });
```

Reply:
Good question.  That might be a little tricky.  `Node` is a Rete class with pre-defined members and `Node` class instances are maintained by the editor.  Instead of making new classes that extend the `Rete.Node` class, we make classe that extend the `Rete.Component` class that manages node creation and processing.  These new `Rete.Component` classes (`rete-node-factory` in our nomenclature) specify what `inputs, `outputs`, `controls`, etc. are added to a node when it is instantiated (the `builder` function) and how a node is updated when the editor processes nodes (the `worker` function).  Nodes also expose `data` which can be modified and accessed anywhere that we have access to a node (which is mainly by asking the editor for a list of nodes).  So the Rete model seems to be to make new classes that extend the `Rete.Component` and these components make use of controls which display and update node data.  I'm not saying we can't extend or change the `Node` class, but rather that it seems to deviate from the Rete model (or perhaps I'm missing something).

Reply:
If it's not obvious, then we should probably leave it for now and maybe a better idea will occur to us down the road.

Reply:
I agree.  This is something we could return to.